File: src/agent_orchestrator/agent_host/signing.py
"""HMAC signing for the agent-host channel.

Each ``tool_call`` issued by the server is bound to a fresh random nonce
and an HMAC-SHA-256 signature over the tuple
``(run_id, tool_call_id, nonce, name)``. The client echoes the same nonce
in ``tool_result`` / ``tool_chunk`` and signs the same tuple — the server
verifies before resolving the pending Future.

Key model
---------

The signing key is **per-session**, minted by the server on accept and
shipped to the client inside the :class:`agent_host.Ack` frame
(``signing_key`` field, 32-byte CSPRNG, hex-encoded). The server stores
the same bytes in :class:`PendingToolCallsRegistry` for the lifetime of
the connection; both sides use the session key for sign and verify.

The dashboard's stable ``JWT_SECRET_KEY`` is **not** sent to the client
(it's a server secret) but is retained as a fallback when no explicit
``key`` is passed — useful for unit tests and for server-internal
signatures that never leave the process.

Threat model
------------

* **Cross-WS tool-result injection.** Without signatures, another
  connection holding a valid CLI JWT could POST ``tool_result`` for a
  tool_call_id it never received. The server already routes results to
  the WS that issued the call; the per-session HMAC adds tamper-evident
  defence in depth and an audit trail.
* **Replay.** ``tool_call_id`` is single-use — once a ``tool_result``
  resolves the pending Future, the call is purged. Replays land on the
  missing-id branch and are rejected.
* **Frame tampering in transit.** Handled by TLS at the transport layer;
  signing is not a substitute.
* **Session-key compromise.** The key dies when the connection dies. A
  client that leaks its key only compromises that one chat session.
  Re-connecting mints a fresh key.
"""

from __future__ import annotations

import hashlib
import hmac
import os
import secrets


class SigningKeyMissingError(RuntimeError):
    """Raised when no signing key can be resolved.

    Triggered when ``compute_signature``/``verify_signature`` is called
    without an explicit ``key`` *and* ``JWT_SECRET_KEY`` is also unset.
    Fail-closed rather than degrade silently to unsigned mode.
    """


def _secret_bytes() -> bytes:
    """Fallback key bytes from ``JWT_SECRET_KEY`` env.

    Used only when the caller does not pass an explicit ``key=`` argument
    (tests and server-internal signatures).
    """
    key = os.environ.get("JWT_SECRET_KEY", "")
    if not key:
        raise SigningKeyMissingError(
            "JWT_SECRET_KEY is required for agent-host signing (or pass an explicit session key)"
        )
    return key.encode("utf-8")


def new_nonce() -> str:
    """16-byte cryptographically-random hex string (32 chars).

    Used once per tool_call; the client must echo it verbatim in the
    matching tool_result / tool_chunk. Reusing the OS CSPRNG keeps the
    project away from custom RNG concerns.
    """
    return secrets.token_hex(16)


def new_session_key() -> bytes:
    """32-byte random session signing key.

    Minted by the server on every agent-host WS accept, shipped to the
    client inside the :class:`agent_host.Ack` frame, and used for HMAC
    sign/verify by both peers for the lifetime of that connection.
    """
    return secrets.token_bytes(32)


def compute_signature(
    *,
    run_id: str,
    tool_call_id: str,
    nonce: str,
    name: str,
    key: bytes | None = None,
) -> str:
    """HMAC-SHA-256 of ``run_id|tool_call_id|nonce|name`` as hex string.

    Pipe-separation is safe because the inputs are server-controlled
    identifiers (UUID hex / opaque names) — no user-supplied content can
    smuggle a pipe into the message. The name is included so an attacker
    who somehow captures a nonce can't repurpose it to a different tool.

    Pass ``key`` to use a session-scoped key (the recommended path —
    the server mints it per connection and ships it inside the ACK
    frame). Omit ``key`` to fall back to ``JWT_SECRET_KEY`` — used by
    server-internal flows and unit tests.
    """
    secret = key if key is not None else _secret_bytes()
    msg = f"{run_id}|{tool_call_id}|{nonce}|{name}".encode()
    return hmac.new(secret, msg, hashlib.sha256).hexdigest()


def verify_signature(
    *,
    run_id: str,
    tool_call_id: str,
    nonce: str,
    name: str,
    signature: str,
    key: bytes | None = None,
) -> bool:
    """Constant-time verification.

    Returns ``False`` on any mismatch and on signatures that fail length
    parsing — never raises. ``hmac.compare_digest`` is mandatory here:
    a naive ``==`` would leak a timing oracle on the first differing byte.
    """
    if not signature:
        return False
    try:
        expected = compute_signature(
            run_id=run_id,
            tool_call_id=tool_call_id,
            nonce=nonce,
            name=name,
            key=key,
        )
    except SigningKeyMissingError:
        # Caller already enforces presence at handshake; treat a late miss
        # as a verification failure (fail-closed) rather than crashing the
        # connection.
        return False
    return hmac.compare_digest(expected, signature)

File: cli/src/agent_host/signing.rs
//! HMAC-SHA-256 signing for the agent-host channel.
//!
//! Mirrors `src/agent_orchestrator/agent_host/signing.py`. The signing
//! key is per-session: minted by the server in [`Ack::signing_key`],
//! decoded once by the client into raw bytes, then used for sign and
//! verify on every `tool_call` / `tool_result` / `tool_chunk`. The
//! dashboard's stable `JWT_SECRET_KEY` is **not** shipped to the
//! client.
//!
//! Threat model
//! ------------
//!
//! * **Cross-WS tool-result injection.** Without signatures, another
//!   connection holding a valid CLI JWT could POST `tool_result` for a
//!   `tool_call_id` it never received. The server already routes results
//!   to the WS that issued the call; the per-session HMAC adds
//!   tamper-evident defence in depth.
//! * **Replay.** `tool_call_id` is single-use server-side.
//! * **Frame tampering in transit.** TLS at the transport layer; signing
//!   is not a substitute.
//! * **Session-key compromise.** The key dies with the connection. A
//!   leaked key only compromises that one chat session.

use hmac::{Hmac, Mac};
use rand::RngCore;
use sha2::Sha256;
use thiserror::Error;

type HmacSha256 = Hmac<Sha256>;

/// 32-byte CSPRNG signing key.
///
/// Server-only helper — clients receive their session key as the
/// `signing_key` hex string in the ACK frame and decode it via
/// [`decode_hex_key`].
pub fn new_session_key() -> [u8; 32] {
    let mut buf = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut buf);
    buf
}

/// 16-byte hex nonce. Used once per tool call.
pub fn new_nonce() -> String {
    let mut buf = [0u8; 16];
    rand::thread_rng().fill_bytes(&mut buf);
    hex::encode(buf)
}

#[derive(Debug, Error)]
pub enum KeyDecodeError {
    #[error("signing_key is empty — server did not issue a session key")]
    Empty,
    #[error("signing_key is not valid hex: {0}")]
    Hex(#[from] hex::FromHexError),
    #[error("signing_key has wrong length: got {0} bytes, expected 32")]
    WrongLength(usize),
}

/// Decode the `Ack.signing_key` hex string to raw bytes.
///
/// Strict on length so a server that ships the wrong size is caught
/// immediately rather than silently producing weak HMACs.
pub fn decode_hex_key(hex_str: &str) -> Result<Vec<u8>, KeyDecodeError> {
    if hex_str.is_empty() {
        return Err(KeyDecodeError::Empty);
    }
    let bytes = hex::decode(hex_str)?;
    if bytes.len() != 32 {
        return Err(KeyDecodeError::WrongLength(bytes.len()));
    }
    Ok(bytes)
}

/// HMAC-SHA-256 of `run_id|tool_call_id|nonce|name` keyed by `key`.
///
/// Returns the hex-encoded digest (64 chars). Pipe-separation is safe
/// because every input is a server-controlled identifier (UUID hex or
/// an opaque tool name); no user-controlled content can smuggle a pipe.
/// `name` is part of the message so a captured nonce can't be reused
/// for a different tool.
pub fn compute_signature(key: &[u8], run_id: &str, tool_call_id: &str, nonce: &str, name: &str) -> String {
    let msg = format!("{run_id}|{tool_call_id}|{nonce}|{name}");
    let mut mac = HmacSha256::new_from_slice(key).expect("HmacSha256 accepts any key length");
    mac.update(msg.as_bytes());
    hex::encode(mac.finalize().into_bytes())
}

/// Constant-time HMAC verification.
///
/// Returns `false` on any mismatch and on signatures that fail hex
/// parsing — never panics. The underlying `MacError` comparison is
/// constant-time by construction; we use it instead of a naive `==`
/// to avoid a timing oracle on the first differing byte.
pub fn verify_signature(
    key: &[u8],
    run_id: &str,
    tool_call_id: &str,
    nonce: &str,
    name: &str,
    signature_hex: &str,
) -> bool {
    if signature_hex.is_empty() {
        return false;
    }
    let provided = match hex::decode(signature_hex) {
        Ok(b) => b,
        Err(_) => return false,
    };
    let msg = format!("{run_id}|{tool_call_id}|{nonce}|{name}");
    let mut mac = match HmacSha256::new_from_slice(key) {
        Ok(m) => m,
        Err(_) => return false,
    };
    mac.update(msg.as_bytes());
    mac.verify_slice(&provided).is_ok()
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    fn key() -> [u8; 32] {
        [42u8; 32]
    }

    #[test]
    fn round_trip() {
        let k = key();
        let sig = compute_signature(&k, "r1", "t1", "n1", "file_write");
        assert!(verify_signature(&k, "r1", "t1", "n1", "file_write", &sig));
    }

    #[test]
    fn empty_signature_rejected() {
        let k = key();
        assert!(!verify_signature(&k, "r", "t", "n", "x", ""));
    }

    #[test]
    fn non_hex_signature_rejected() {
        let k = key();
        assert!(!verify_signature(&k, "r", "t", "n", "x", "zzzz"));
    }

    #[test]
    fn tampered_run_id_rejected() {
        let k = key();
        let sig = compute_signature(&k, "r1", "t1", "n1", "file_write");
        assert!(!verify_signature(&k, "r2", "t1", "n1", "file_write", &sig));
    }

    #[test]
    fn tampered_tool_call_id_rejected() {
        let k = key();
        let sig = compute_signature(&k, "r1", "t1", "n1", "file_write");
        assert!(!verify_signature(&k, "r1", "t2", "n1", "file_write", &sig));
    }

    #[test]
    fn tampered_nonce_rejected() {
        let k = key();
        let sig = compute_signature(&k, "r1", "t1", "n1", "file_write");
        assert!(!verify_signature(&k, "r1", "t1", "n2", "file_write", &sig));
    }

    #[test]
    fn tampered_name_rejected() {
        let k = key();
        let sig = compute_signature(&k, "r1", "t1", "n1", "file_write");
        assert!(!verify_signature(&k, "r1", "t1", "n1", "file_read", &sig));
    }

    #[test]
    fn wrong_key_rejected() {
        let sig = compute_signature(&key(), "r", "t", "n", "x");
        let other = [0u8; 32];
        assert!(!verify_signature(&other, "r", "t", "n", "x", &sig));
    }

    #[test]
    fn signature_is_hex_64() {
        let k = key();
        let sig = compute_signature(&k, "r", "t", "n", "x");
        assert_eq!(sig.len(), 64);
        assert!(sig.chars().all(|c| c.is_ascii_hexdigit()));
    }

    #[test]
    fn new_session_key_random() {
        let a = new_session_key();
        let b = new_session_key();
        assert_ne!(a, b);
        assert_eq!(a.len(), 32);
    }

    #[test]
    fn new_nonce_random_hex_32() {
        let a = new_nonce();
        let b = new_nonce();
        assert_ne!(a, b);
        assert_eq!(a.len(), 32);
        assert!(a.chars().all(|c| c.is_ascii_hexdigit()));
    }

    #[test]
    fn decode_hex_key_strict() {
        let k = new_session_key();
        let h = hex::encode(k);
        let decoded = decode_hex_key(&h).unwrap();
        assert_eq!(decoded, k.to_vec());

        assert!(matches!(decode_hex_key(""), Err(KeyDecodeError::Empty)));
        assert!(matches!(
            decode_hex_key("abcd"),
            Err(KeyDecodeError::WrongLength(2))
        ));
        assert!(matches!(decode_hex_key("zz"), Err(KeyDecodeError::Hex(_))));
    }

    #[test]
    fn python_compat_known_vector() {
        // Cross-check against the Python implementation: with key="x"*32
        // (the unit-test fixture in tests/test_agent_host_protocol.py),
        // run_id="r", tool_call_id="t", nonce="n", name="x" must
        // produce a deterministic digest.  Compute it locally and
        // assert it matches what the Python tests assert. This is a
        // self-pinned vector — if either implementation drifts the
        // suite catches it.
        let k = b"x".repeat(32);
        let sig = compute_signature(&k, "r", "t", "n", "x");
        // Verifying through verify_signature ensures we have NOT
        // silently changed the message-construction format.
        assert!(verify_signature(&k, "r", "t", "n", "x", &sig));
        // And tampering with any field still fails.
        assert!(!verify_signature(&k, "r", "t", "n", "y", &sig));
    }
}
